File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/handler/AnimationHandlerImpl.java
package com.github.nickpakhomov.hotel_viewer.handler;

import android.content.Context;
import android.os.Handler;
import android.view.MenuItem;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;

import com.github.nickpakhomov.hotel_viewer.ui.activities.DetailActivity;
import com.github.nickpakhomov.hotel_viewer.ui.activities.MainActivity;
import com.github.nickpakhomov.hotel_viewer.R;

/**
 * Class that handles "refresh" icon animation
 * Created by Nikolay Pakhomov on 28/08/17.
 */

public class AnimationHandlerImpl implements AnimationHandler {
    private static final int ANIMATION_STOP_DELAY = 500; //ms
    
    private MenuItem mItem;
    private Context mContext;
    private FetcherHandlerImpl mFetcherHandler;
    private ImageView mRefreshButton;
    private Handler mAnimationStopHandler;
    
    public AnimationHandlerImpl(MenuItem item, Context context) {
        mItem = item;
        mContext = context;
        
        initAnimationHandler();
    }
    
    public void setFetcher(FetcherHandlerImpl mFetcherHandler) {
        this.mFetcherHandler = mFetcherHandler;
    }
    
    private void initAnimationHandler() {
        mItem.setActionView(R.layout.iv_refresh);
        mRefreshButton = (ImageView) mItem.getActionView().findViewById(R.id.refreshButton);
        mRefreshButton.setOnClickListener(v -> update());
    }
    
    private void update() {
        stopAnimationImmediately();
        
        if (mContext instanceof MainActivity)
            mFetcherHandler.fetchHotelPreviewsData();
        else if (mContext instanceof DetailActivity)
            mFetcherHandler.fetchHotelsData(mFetcherHandler.getLastHotelID());
    }
    
    public void startAnimation() {
        mRefreshButton.setImageResource(R.drawable.ic_sync_arrows);
        Animation rotation = AnimationUtils.loadAnimation(mContext, R.anim.refresh_rotation);
        rotation.setFillAfter(true);
        mRefreshButton.startAnimation(rotation);
    }
    
    private void stopAnimationImmediately() {
        if (mRefreshButton != null)
            mRefreshButton.clearAnimation();
    }
    
    /**
     * Extending animation to extra 500ms for UX
     */
    public void stopAnimation() {
        mAnimationStopHandler =  new Handler();
        mAnimationStopHandler.
                postDelayed(() -> mRefreshButton.clearAnimation(), ANIMATION_STOP_DELAY);
    }
    
    public void setStatusError() {
        stopAnimation();
        mRefreshButton.setImageResource(R.drawable.ic_sync_arrows_error);
    }
    
    public void onDestroy() {
        mItem = null;
        mContext = null;
        if (mAnimationStopHandler != null) {
            mAnimationStopHandler.removeCallbacksAndMessages(null);
            mAnimationStopHandler = null;
        }
        stopAnimationImmediately();
        mRefreshButton = null;
    }
}

File: app/src/main/java/com/github/nickpakhomov/hotel_viewer/handler/AnimationHandlerImpl.kt
package com.github.nickpakhomov.hotel_viewer.handler

import android.content.Context
import android.os.Handler
import android.view.MenuItem
import android.view.animation.AnimationUtils
import android.widget.ImageView
import com.github.nickpakhomov.hotel_viewer.R
import com.github.nickpakhomov.hotel_viewer.ui.activities.DetailActivity
import com.github.nickpakhomov.hotel_viewer.ui.activities.MainActivity

/**
 * Class that handles "refresh" icon animation
 * Created by Nikolay Pakhomov on 28/08/17.
 */

class AnimationHandlerImpl(
        mItem: MenuItem?,
        private var mContext: Context) : AnimationHandler {

    private lateinit var mFetcherHandler: FetcherHandler
    private var mRefreshButton: ImageView
    private lateinit var mAnimationStopHandler: Handler

    init {
        mItem?.setActionView(R.layout.iv_refresh)
        mRefreshButton = mItem?.actionView?.findViewById(R.id.refreshButton) as ImageView
        mRefreshButton.setOnClickListener { update() }
    }

    override fun setFetcher(mFetcherHandler: FetcherHandler) {
        this.mFetcherHandler = mFetcherHandler
    }

    private fun update() {
        stopAnimationImmediately()

        if (mContext is MainActivity)
            mFetcherHandler.fetchHotelPreviewsData()
        else if (mContext is DetailActivity)
            mFetcherHandler.fetchHotelsData(mFetcherHandler.getLastHotelID())
    }

    override fun startAnimation() {
        mRefreshButton.setImageResource(R.drawable.ic_sync_arrows)
        val rotation = AnimationUtils.loadAnimation(mContext, R.anim.refresh_rotation)
        rotation.fillAfter = true
        mRefreshButton.startAnimation(rotation)
    }

    private fun stopAnimationImmediately() {
        mRefreshButton.clearAnimation()
    }

    /**
     * Extending animation to extra 500ms for UX
     */
    override fun stopAnimation() {
        mAnimationStopHandler = Handler()
        mAnimationStopHandler.postDelayed({ mRefreshButton.clearAnimation() },
                ANIMATION_STOP_DELAY.toLong())
    }

    override fun setStatusError() {
        stopAnimation()
        mRefreshButton.setImageResource(R.drawable.ic_sync_arrows_error)
    }

    override fun onDestroy() {
        mAnimationStopHandler.removeCallbacksAndMessages(null)
        stopAnimationImmediately()
    }

    companion object {
        private val ANIMATION_STOP_DELAY = 500 //ms
    }
}
